petab/migrations.py

```diff
@@ -0,0 +1,116 @@
+"""Functions for migrating PEtab files from different versions"""
+
+import logging
+import pandas as pd
+
+from . import Problem
+from . import get_notnull_columns
+from .C import *
+from .lint import lint_problem
+from .measurements import get_placeholders
+from .core import get_observable_id
+from .sbml import get_sigmas, get_observables
+
+
+def sbml_observables_to_table(problem: Problem):
+    """Transform PEtab files where observables are defined inside the SBML
+    model to the newer format where they are specified in a separate table.
+
+    For details see https://github.com/ICB-DCM/PEtab/issues/241.
+
+    Modifies ``problem`` in place. Does not alter any files.
+
+    Assumes the input is adheres to the PEtab format. You may want to check
+    that with ``petablint`` from PEtab 0.0.2.
+    """
+
+    logging.basicConfig(level=logging.DEBUG)
+
+    # Ensure all measurements for the same observable have the same
+    # observableTransformation and noiseModel
+
+    measurement_df = problem.measurement_df
+    grouping_cols = get_notnull_columns(
+        measurement_df, [OBSERVABLE_ID, OBSERVABLE_TRANSFORMATION,
+                         NOISE_DISTRIBUTION])
+    grouped = measurement_df.groupby(grouping_cols).count().reset_index()
+
+    if len(grouped[OBSERVABLE_ID]) != len(grouped[OBSERVABLE_ID]):
+        raise ValueError("In order to create an observable table, all "
+                         "measurements for the same observable must have the "
+                         "same observableTransformation and noiseModel."
+                         "Multiple observableTransformation and noiseModel "
+                         "has to be handled by different observables.")
+
+    # get observables and sigmas from SBML file and directly remove them
+    observables = get_observables(problem.sbml_model, remove=True)
+    sigmas = get_sigmas(problem.sbml_model, remove=True)
+    assert observables.keys() == sigmas.keys()
+
+    if not observables:
+        raise RuntimeError("No observables in SBML model to convert. "
+                           "Has this model already been processed?")
+
+    # Create observable dataframe and add to `problem`
+
+    for obs_id, noise in sigmas.items():
+        observables[obs_id][NOISE_FORMULA] = noise
+
+    # set observableTransformation and noiseModel
+    for obs_id_long in observables.keys():
+        obs_id = get_observable_id(obs_id_long)
+        cur_mes_df = measurement_df[measurement_df[OBSERVABLE_ID] == obs_id]
+        if not len(cur_mes_df):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not use `len(SEQUENCE)` without comparison to determine if a sequence is empty](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4885513)